lib/remi/data_subjects/gsheet_file.rb

```diff
@@ -0,0 +1,163 @@
+    attr_reader :itk_file
+    attr_reader :data
+
+
+    def authorize()
+
+      FileUtils.mkdir_p(File.dirname(@credentials_path))
+      client_id   = Google::Auth::ClientId.from_file(@client_secrets_path)
+      token_store = Google::Auth::Stores::FileTokenStore.new(file: @credentials_path)
+      authorizer  = Google::Auth::UserAuthorizer.new(client_id, [Google::Apis::DriveV3::AUTH_DRIVE, Google::Apis::SheetsV4::AUTH_SPREADSHEETS], token_store)
+      user_id     = 'default'
+      creds       = authorizer.get_credentials(user_id)
+      if creds.nil?
+        raise "Error, Credentials file does not exist"
+      end
+      creds
+    end
+
+
+    def get_file_list(folder_id)
+
+      service                                 = Google::Apis::DriveV3::DriveService.new
+      service.client_options.application_name = @application_name
+      service.authorization                   = authorize()
+      response = service_list_files(service, folder_id)
+      response.files
+
+    end
+
+    def service_list_files(service, folder_id)
+      service.list_files(q: "'#{folder_id}' in parents", page_size: 10, order_by: 'createdTime desc', fields: 'nextPageToken, files(id, name, createdTime, mimeType)')
+    end
+
+    def get_spreadsheet_vals(service, spreadsheet_id)
+      service.get_spreadsheet_values(spreadsheet_id, 'Sheet1')
+    end
+
+    def extract()
+
+      service                                 = Google::Apis::SheetsV4::SheetsService.new
+      service.client_options.application_name = @application_name
+      service.authorization                   = authorize()
+      @data                                   = []
+      file_list  = get_file_list(@default_folder_id)
+
+      file_list.each do |file|
+        file = file.to_h
+        response = get_spreadsheet_vals(service, file[:id])
+        data.push(response)
+      end
+
+      self
+
+    end
+
+    # @return [Array<Extractor::FileSystemEntry>] (Memoized) list of objects in the bucket/prefix
+    def all_entries
+      @all_entries ||= all_entries!
+    end
+
+    # @return [Array<Extractor::FileSystemEntry>] (Memoized) list of objects in the bucket/prefix
+    def all_entries!
+
+      gsheet_entries = get_file_list(@default_folder_id)
+      gsheet_entries.map do |entry|
+        entry = entry.to_h
+        FileSystemEntry.new(
+          pathname:       File.join(@default_folder_id, entry[:name]),
+          create_time:    entry[:created_time],
+          modified_time:  entry[:created_time]
+        )
+      end
+    end
+
+    private def init_gsheet_extractor(*args, credentials:, **kargs)
+
+      @default_folder_id   = credentials.fetch(:folder_id)
+      @oob_uri             = 'urn:ietf:wg:oauth:2.0:oob'
+      @application_name    = 'ITK-ETL'
+
+      @client_secrets_path = File.join(
+        Dir.home,
+        '.credentials/client_secret.json'
+      )
+      @credentials_path = File.join(
+        Dir.home,
+        '.credentials/sheets.googleapis.com-ruby-itk.yaml'
```

Comment:
that is where the file gets stored, as you can see, it writes that file then uses it in the authorization step. it could be but I dont really know if there is a reason to make it configurable? we take all the parameters from the internet then we write the necessary files